core/include/scipp/core/element/rebin.h

```diff
@@ -37,13 +38,29 @@ static constexpr auto rebin = overloaded{
           const auto delta = std::min<double>(xn_high, xo_high) -
                              std::max<double>(xn_low, xo_low);
           const auto owidth = xo_high - xo_low;
-          const auto scale = delta / owidth;
-          if constexpr (is_ValueAndVariance_v<
-                            std::decay_t<decltype(data_old)>>) {
+
+          constexpr bool isValAndVariance =
+              is_ValueAndVariance_v<std::decay_t<decltype(data_old)>>;
+
+          if constexpr (isValAndVariance) {
+            // Avoid implicit promotion and demotion when multiplying by scale
+            const auto scale =
+                std::remove_reference_t<decltype(data_new.value[0])>(delta /
+                                                                     owidth);
             data_new.value[inew] += data_old.value[iold] * scale;
             data_new.variance[inew] += data_old.variance[iold] * scale;
           } else {
-            data_new[inew] += data_old[iold] * scale;
+            if constexpr (!isValAndVariance &&
```

Comment:
The compiler attempts to eval this condition for Val and Variance types too regardless of the else branch

Without `isValAndVariance` to short circuit the condition compilation fails since there is no `operator[]`

Reply:
Why did it compile before? What is the difference?

Reply:
This condition `std::is_same_v<std::remove_reference_t<decltype(data_new[0])>, bool>` didn't exist since we didn't have to have special `bool`/`T` handling

decltype fails without the short circuit as it tried to do `data_new[0]` on VarAndVariance types

Reply:
Use something like
```cpp
using ElemT = std::decay_t<std::conditional_t<isValAndVariance, decltype(data_new.value[0], decltype(data_new[0])>>
```
somewhere. This should also solve the comment below.

Reply:
No luck, `decltype` is evaluated before `conditional_t` following the evaluation rules

If we could hoist it up somehow to something like this I reckon it would work
`decltype(std::conditional_t<isValAndVariance, data_new.value[0], data_new[0]>)`
But conditional expects types, so were back to a chicken and egg problem.

I also tried using the short circuit property of ternarys:
`decltype(isValAndVariance ? data_new.value[0] : data_new[0])` but the LHS and RHS of the operator must be both evaluable at compile time 

I'm sure there is a really simple pattern or STL container were missing here which will make this easy :thinking:

Reply:
I suppose you need a helper struct with specialization. Grep for `element_type`, it should be defined somewhere, but might not include this case (but you can easily make an equivalent).